### PR TITLE
Turn off hand rays only when hand is near a grabbable object

### PR DESCRIPTION
**Pull request by julenka**
## Overview
If the hand is near the spatial mesh (SR), hand rays should not turn off. In fact, rays should only turn off if the hand is near a grabbable or touchable thing. This PR updates SpherePointer to return `IsNearObject` and `IsInteractionEnabled` only when the hand is near a collider that has a NearInteractionGrabbable on it. It also adds Grab Layer Masks, which allows developers to further prioritize and filter which objects a pointer can interact with. 

After this change, an object must both be on a Grabbable Layer as well as have a NearInteractionGrabbable component in order for hand rays to turn off. The Grabbable Layer is by default set to everything **except** Ignore Raycast and Spatial Awareness.

Fixes: #5748, #4203, #5621 

## Changes
- Add a Grabbable Layer Mask to Sphere Pointer to allow devs to easily disable all grabbable objects, and to improve performance
- The default Grabbale Layer Mask for the grab pointer is everything but Ignore Raycast and Spatial Awareness so that sphere casts don't check the spatial mesh (for perf).
- I also am caching the sphere query results to ensure we only do 3 sphere casts per frame (down from 11 without caching).
-  I snuck in a fix to pointer locking -- previously a pointer would lock when you clicked on any collidable, even if the collidable did not actually have any pointer handlers on it, or above it. now, the pointer will only lock if the pointer is actually grabbing an object that would handle pointer events. I may actually end up pulling that change into a separate PR.
- Add tests to verify my fix, and documentation reflecting the new grabbable behavior
- I also fixed the sphere pointer inspector -- it was actually not displaying several sphere pointer fields that it should have
- I removed the DebugSphere because it was not very useful, added a bit of extra code, and had bugs (did not turn off when debugsphere == false).

## Verification
- All playmode tests pass, new tests pass
- Verified that my new tests would have failed in old code
- Verified that my change actually fixes the bug (ray no longer hides when near SR)
- Runs on HoloLens 2 with no adverse perf impact.

### CODE REVIEW

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointer.cs [hunk 1]
```diff
@@ -88,20 +116,21 @@ public override void OnPreSceneQuery()
...
+                var toQuery = new SpherePointerQueryInfo[] { queryBufferNearObjectRadius, queryBufferInteractionRadius };
+                for (int j = 0; j < toQuery.Length; j++)
                 {
-                    if (debugSphere == null)
+                    for (int i = 0; i < layerMasks.Length; i++)
                     {
-                        debugSphere = GameObject.CreatePrimitive(PrimitiveType.Sphere).transform;
-                        debugSphere.localScale = Vector3.one * SphereCastRadius * 2;
-                        Destroy(debugSphere.gameObject.GetComponent<Collider>());
+                        if (toQuery[j].TryUpdateQueryBufferForLayerMask(layerMasks[i], pointerPosition, triggerInteraction))
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointer.cs]: I think a for loop over toQuery here is excessive, remove

**Comment by davidkline-ms** [comment]
> After this change, an object must both be on a Grabbable Layer as well as have a NearInteractionGrabbable component

why the layer requirement? can the same be accomplished by just checking for NearInteractionGrabbable/Touchable?

**Comment by julenka** [comment]
> > After this change, an object must both be on a Grabbable Layer as well as have a NearInteractionGrabbable component
> 
> why the layer requirement? can the same be accomplished by just checking for NearInteractionGrabbable/Touchable?

The default Grabbale Layer Mask for the grab pointer is everything but Ignore Raycast and Spatial Awareness so that sphere casts don't check the spatial mesh (for perf). The poke pointer already uses this approach and I wanted to mimic it. Additionally, this will make it easy for devs to disable large collections of UI by just changing layers without needing to go down and remove grabbables from every component.

**Comment by Troy-Ferrell** [comment]
FYI @julenka I believe this also fixes #5621

File: Documentation/Input/Pointers.md [hunk 2]
```diff
@@ -90,6 +89,7 @@ Useful Poke Pointer properties:
 - *TouchableDistance*: Maximum distance a which a touchable surface can be interacted with
 - *Visuals*: Game object used to render finger tip visual (the ring on finger, by default).
 - *Line*: Optional line to draw from fingertip to the active input surface.
+- *Poke Layer Masks* - A prioritized array of LayerMasks to determine which possible GameObjects a the pointer can interact with and the order of interaction to attempt. Note that a GameObject must also have a `NearInteractionTouchable` component in order to interact with a poke pointer.
```
Comment (thalbern) [inline Documentation/Input/Pointers.md]:
```suggestion
- *Poke Layer Masks* - A prioritized array of LayerMasks to determine which possible GameObjects the pointer can interact with and the order of interaction to attempt. Note that a GameObject must also have a `NearInteractionTouchable` component in order to interact with a poke pointer.
```

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointer.cs [hunk 3]
```diff
@@ -16,7 +17,7 @@ public class SpherePointer : BaseControllerPointer, IMixedRealityNearPointer
 
         [SerializeField]
         [Min(0.0f)]
-        [Tooltip("Additional distance between SphereCastRadius and NearObjectRadius")]
+        [Tooltip("Additional distance on top of sphere cast radius when pointer is considered 'near' an object and far interaction will turn off")]
         private float nearObjectMargin = 0.2f;
         /// <summary>
         /// Additional distance between <see cref="BaseControllerPointer.SphereCastRadius"/> and <see cref="NearObjectRadius"/>.
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointer.cs]:
```suggestion
        /// Additional distance on top of<see cref="BaseControllerPointer.SphereCastRadius"/> when pointer is considered 'near' an object and far interaction will turn off.
```

**Comment by julenka** [comment]
Completing PR. All checks passed with my code change. Re-ran only docs test because I only made changes to comments.

**Comment by davidkline-ms** [comment]
@julenka, is this a 2.2.0 fix? If so, I can cherry-pick it into the stabilization branch.

**Comment by keveleigh** [comment]
@davidkline-ms This was in the 2.1 release: 
image

**Pull request merged**